### PR TITLE
«drupal-project integration proof of concept»

### PR DESCRIPTION
**Pull request by novia713**
just for toying,
this is a proof of concept on how we could integrate «drupal-project» with Drupal Console.

this way instead of downloading the module, the program just executes a 
«_composer require drupal/module_»

i know the code is horrible, but just for giving the idea to more wise than me people ^^

seleccion_516

### CODE REVIEW

File: src/Command/Module/DownloadCommand.php [hunk 1]
```diff
@@ -40,6 +41,12 @@ protected function configure()
                 '',
                 InputOption::VALUE_NONE,
                 $this->trans('commands.module.download.options.latest')
+            )
+            ->addOption(
+                'druproj',
```
Comment (darol100) [inline src/Command/Module/DownloadCommand.php]: @novia713 Can we rename `druproj ` to `composer` ?

Comment (novia713) [inline src/Command/Module/DownloadCommand.php]: @darol100 renamed, sir

**Comment by novia713** [comment]
i added initial support for installing a site with Composer with «site:new»
it's just an idea, there are many improvements to do
just brainstorming ... !

**Comment by novia713** [comment]
**_site:new --composer_ & _module:download --composer_ should work both.**

@TODO: formatting code :tada: 
@TODO: themes :balloon: 

--
 - not sure if for installing a module with composer should be «_module:download --composer_» or «_module:install --composer_» anyway, now it works  just in «_md_»
 - _--latest_ option works well in «_sn_», also works but in «_md_» is not recommended because there are not stable modules for D8 yet
- in both cases always appears a question for if «_stable_» versions is preferred or not. also, use with caution in «_md_», given the actual D8 modules context

--

i have an extrange problem when downloading modules with Composer.
Right now, it usually gets an error, as shown below:

seleccion_537


anyway, all the logic is quite stable, except the crucial command:
`$cmd = 'composer require "drupal/' . $module .':' . $version . '"';`
i don't know, perhaps the command is wrong or maybe related with the «minimum stability» of the project composer.json. :disappointed: 

feedback is welcome! :8ball:

**Comment by novia713** [comment]
```
"repositories": [
    {
        "type": "composer",
        "url": "https://packagist.drupal-composer.org"
    }
],
```
must be added to the project composer.json for downloading the modules via «_module:download --composer_»
i'm working on it

**Comment by novia713** [comment]
question: should we remove from modules releases option list all related items with D7?

**Comment by novia713** [comment]
another doubt: i fear this line is not working well:
https://github.com/hechoendrupal/DrupalConsole/pull/2207/commits/387431a0fe3e483272032a6e07571a8629298cb3#diff-1024415cb7edf3e9fcf7a3bb19b88463R25

perhaps would be better:
`$rootPath = $this->getApplication()->getSite()->getSiteRoot()`

**Comment by esolitos** [comment]
Nice, +1 to this.

**Comment by novia713** [comment]
//@TODO: add support for https://github.com/drupal-composer/drupal-project#how-can-i-apply-patches-to-downloaded-modules

the idea would be 2 commands:
 - 1 command given the url of the patch and to write the "extra" json in the project's composer.json from DrupalConsole
- another command for apply that patch from DrupalConsole

**Comment by novia713** [comment]
//@TODO: handle guzzle exceptions when no module name found to download (using --composer)
seleccion_540

**Comment by jmolivas** [comment]
@novia713 This is looking awesome 👍

**Comment by novia713** [comment]
merged with latest changes from upstream :dart:

**Comment by novia713** [comment]
catching Guzzle errors if no module releases found

seleccion_580

**Comment by novia713** [comment]
«module:uninstall --composer» removes an array of modules in composer.json

**Comment by novia713** [comment]
i would merge this together with #2268 
otherwise «uninstall» command will fail, because _modulesUninstallQuestion()_ is in **src/Command/ProjectDownloadTrait.php** and this is added in #2268 

anyway, if the merge is very conflictive i think i'm able to make a new PR with both previous PRs joined if it were necessary

**Comment by jmolivas** [comment]
@novia713  I like the idea of merging this with #2268 sync and fix conflicts on that one and close this after that I will be able to top merge #2268.

**Comment by novia713** [comment]
i merged this PR in #2268 
i close this one

**Pull request closed without merging**